core/shared/src/main/scala/zio/Config.scala

```diff
@@ -154,11 +154,16 @@ object Config {
   final class Secret private (private val raw: Array[Char]) { self =>
     override def equals(that: Any): Boolean =
       that match {
-        case that: Secret =>
-          self.raw.length == that.raw.length &&
-            (0 until raw.length).foldLeft(true) { (b, i) =>
-              self.raw(i) == that.raw(i) && b
-            }
+        case that: Secret => {
+          val selfLength = self.raw.length
+          val thatLength = that.raw.length
+          var isEqual    = if (selfLength == thatLength) 0 else 1
+          for (i <- 0 until selfLength) {
```

Comment:
This this might seem like a loop, but it actually compiles to `(0 until selfLength).foreach`. Because of this, `var isEqual` is compiled to a heap variable (similar to `AtomicReference`) so that it can be used in the `Function1` of `foreach`.

I think we need to rewrite this as a while-loop instead